Functions/Assertions/Should.ps1

```diff
@@ -112,36 +166,70 @@ function Invoke-LegacyAssertion($assertionEntry, $shouldArgs, $valueToTest, $fil
     #     }
     # )
 
-    $negate = -not $shouldArgs.PositiveAssertion
+    # $negate = -not $shouldArgs.PositiveAssertion
 
-    $testResult = (& $assertionEntry.Test $valueToTest $shouldArgs.ExpectedValue -Negate:$negate)
-    if (-not $testResult.Succeeded) {
-        throw ( New-ShouldErrorRecord -Message $testResult.FailureMessage -File $file -Line $lineNumber -LineText $lineText )
-    }
+    # $testResult = (& $assertionEntry.Test $valueToTest $shouldArgs.ExpectedValue -Negate:$negate)
+    # if (-not $testResult.Succeeded) {
+    #     throw ( New-ShouldErrorRecord -Message $testResult.FailureMessage -File $file -Line $lineNumber -LineText $lineText )
+    # }
 }
 
 function Invoke-Assertion {
+    [CmdletBinding()]
     param (
-        [object] $AssertionEntry,
-        [System.Collections.IDictionary] $BoundParameters,
-        [object] $valuetoTest,
-        [string] $File,
-        [int] $LineNumber,
-        [string] $LineText,
-        [switch] $Negate,
-        [Management.Automation.SessionState] $CallerSessionState
+        [Parameter(Mandatory)]
+        [ValidateNotNull()]
+        [object]
+        $AssertionEntry,
+
+        [Parameter(Mandatory)]
+        [System.Collections.IDictionary]
+        $BoundParameters,
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string]
+        $File,
+
+        [Parameter(Mandatory)]
+        [int]
+        $LineNumber,
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string]
+        $LineText,
+
+        [Parameter(Mandatory)]
+        [Management.Automation.SessionState]
+        $CallerSessionState,
+
+        [Parameter()]
+        [switch]
+        $Negate,
+
+        [Parameter()]
+        [AllowNull()]
+        [object]
+        $ValueToTest
     )
-
-    $testResult = & $AssertionEntry.Test -ActualValue $valuetoTest -Negate:$Negate -CallerSessionState $CallerSessionState @BoundParameters
-    if (-not $testResult.Succeeded) {
-        throw ( New-ShouldErrorRecord -Message $testResult.FailureMessage -File $file -Line $lineNumber -LineText $lineText )
-    }
-    else {
-        #extract data to return if there are any on the object
-        $data = $testResult.psObject.Properties.Item('Data')
-        if ($data) {
-            $data.Value
+    try {
+        # Write-Host "Bound Parameters: $($PSBoundParameters | Out-String)"
+        $testResult = & $AssertionEntry.Test -ActualValue $ValueToTest -Negate:$Negate -CallerSessionState $CallerSessionState @BoundParameters
+        # Write-Host "Test result: $testResult"
+        if (-not $testResult.Succeeded) {
+            New-ShouldErrorRecord -Message $testResult.FailureMessage -File $file -Line $lineNumber -LineText $lineText
```

Comment:
This is imho where there "magic" happens. With `if ($shouldThrow) { throw $shouldErrorRecord } else { $currentTest.ErrorRecords.Add($shouldErrorRecord) }`. And that is about it. At least that is how I see it with my naive view right now :)